pkg/autoscaling/estimator/percentile.go

```diff
@@ -66,28 +65,71 @@ func (e *PercentileResourceEstimator) GetResourceEstimation(evpa *autoscalingapi
 				Namespace:     evpa.Namespace,
 				WorkloadName:  evpa.Spec.TargetRef.Name,
 				ContainerName: containerName,
-				Selector:      labels.Everything(),
+				Selector:      selector,
 			},
 		},
 	}
-
 	memConfig := getMemConfig(config)
-	tsList, err = utils.QueryPredictedValues(e.Predictor, caller, memConfig, memoryMetricNamer)
+
+	var errs []error
+	// first register cpu & memory, or the memory will be not registered before the cpu prediction succeed
+	err1 := e.Predictor.WithQuery(cpuMetricNamer, caller, *cpuConfig)
+	if err1 != nil {
+		errs = append(errs, err1)
+	}
+	err2 := e.Predictor.WithQuery(memoryMetricNamer, caller, *memConfig)
+	if err2 != nil {
+		errs = append(errs, err2)
+	}
+	if len(errs) > 0 {
+		return nil, fmt.Errorf("failed to register metricNamer: %v", errs)
+	}
+
+	var predictErrs []error
+	var noValueErrs []error
+	tsList, err := e.Predictor.QueryRealtimePredictedValues(context.TODO(), cpuMetricNamer)
+	if err != nil {
+		predictErrs = append(predictErrs, err)
+	}
+
+	if len(tsList) > 1 && len(tsList[0].Samples) > 1 {
+		cpuValue := int64(tsList[0].Samples[0].Value * 1000)
+		recommendResource[corev1.ResourceCPU] = *resource.NewMilliQuantity(cpuValue, resource.DecimalSI)
+	} else {
+		noValueErrs = append(noValueErrs, fmt.Errorf("no value retured for queryExpr: %s", cpuMetricNamer.BuildUniqueKey()))
+	}
+
+	tsList, err = e.Predictor.QueryRealtimePredictedValues(context.TODO(), memoryMetricNamer)
 	if err != nil {
-		return nil, err
+		predictErrs = append(predictErrs, err)
 	}
 
-	if len(tsList) < 1 || len(tsList[0].Samples) < 1 {
-		return nil, fmt.Errorf("no value retured for queryExpr: %s", memoryMetricNamer.BuildUniqueKey())
+	if len(tsList) > 1 && len(tsList[0].Samples) > 1 {
+		memValue := int64(tsList[0].Samples[0].Value)
+		recommendResource[corev1.ResourceMemory] = *resource.NewQuantity(memValue, resource.BinarySI)
+	} else {
+		noValueErrs = append(noValueErrs, fmt.Errorf("no value retured for queryExpr: %s", memoryMetricNamer.BuildUniqueKey()))
 	}
 
-	memValue := int64(tsList[0].Samples[0].Value)
-	recommendResource[corev1.ResourceMemory] = *resource.NewQuantity(memValue, resource.BinarySI)
+	// all failed
+	if len(recommendResource) == 0 {
+		return recommendResource, fmt.Errorf("all resource predicted failed, predictErrs: %v, noValueErrs: %v", predictErrs, noValueErrs)
+	}
 
+	// at least one succeed
 	return recommendResource, nil
 }
 
 func (e *PercentileResourceEstimator) DeleteEstimation(evpa *autoscalingapi.EffectiveVerticalPodAutoscaler) {
+	selector, err := e.TargetFetcher.Fetch(&corev1.ObjectReference{
```

Comment:
Is this selector only needed for metric server or common for both prom and metric server ?

Reply:
> Is this selector only needed for metric server or common for both prom and metric server ?

maybe for prom is useful too, because we select containers and pods name by regex in prometheus now, but maybe use kube -state-metrics  is more safe for selecting pods by kube_pod_labels

metrics server is only can get pods metrics and node metrics for default metric server, so if we want to get workload or container metrics, we must has label selector to list pods, or we will list all pods metrics each time add sample.

you can see line `pkg/providers/metricserver/rest_metric_client.go:156`